apps/synthwave/app.js

```diff
@@ -0,0 +1,760 @@
+};
+
+const unsigned int terrainLength = 12;
+const unsigned int terrainWidth = 12;
+unsigned char terrain[terrainLength][terrainWidth];
+unsigned int _rngState;
+unsigned int rng() {
+    _rngState ^= _rngState << 17;
+    _rngState ^= _rngState >> 13;
+    _rngState ^= _rngState << 5;
+    return _rngState;
+}
+
+void shiftTerrain() {
+  for (int i = terrainLength - 1; i > 0; --i) {
+    for (int x = 0; x < terrainWidth; ++x)
+      terrain[i][x] = terrain[i-1][x];
+  }
+
+  for (int x = 0; x < terrainWidth; ++x)
+    terrain[0][x] = (int(terrain[0][x]) + ((rng() & 0x7F) + 0x7)) >> 1;
+  int mid = terrainWidth >> 1;
+  terrain[0][mid-1] >>= 1;
+  terrain[0][mid  ] = 0;
+  terrain[0][mid+1] = 0;
+  terrain[0][mid+2] = 0;
+  terrain[0][mid+3] >>= 1;
+}
+
+void init(unsigned char* _fb, int _stride, unsigned char* _sint) {
+  fb = _fb;
+  stride = _stride;
+  sint = _sint;
+  _rngState = 1013904223;
+  for (int i = 0; i < terrainLength; ++i)
+    shiftTerrain();
+  speed.x = 0;
+  speed.y = 0;
+  speed.z = 0;
+  position.x = 100 << 8;
+  position.y = -150 << 8;
+  position.z = 100 << 8;
+  rotation.x = 0;
+  rotation.y = 256 << 8;
+  rotation.z = 0;
+  scale.x = 1 << 8;
+  scale.y = 1 << 8;
+  scale.z = 1 << 8;
+}
+
+int sin(int angle) {
+  int a = (angle >> 7) & 0xFF;
+  if (angle & (1 << 15))
+    a = 0xFF - a;
+  int v = sint[a];
+  if (angle & (1 << 16))
+    v = -v;
+  return v;
+}
+
+int cos(int angle) {
+  return sin(angle + 0x8000);
+}
+
+void setCamera(int x, int y, int z) {
+  camera.x = x;
+  camera.y = y;
+  camera.z = z;
+}
+
+unsigned int solid(unsigned int c) {
+  c &= 7;
+  c |= c << 3;
+  c |= c << 6;
+  c |= c << 12;
+  c |= c << 24;
+  return c;
+}
+
+unsigned int alternate(unsigned int a, unsigned int b) {
+  unsigned int c = (a & 7) | ((b & 7) << 3);
+  c |= c << 6;
+  c |= c << 12;
+  c |= c << 24;
+  return c;
+}
+
+void drawHLine(int x, unsigned int y, int l, unsigned int c) {
+  if (x < 0) {
+    l += x;
+    x = 0;
+  }
+  if (x + l >= 176) {
+    l = 176 - x;
+  }
+  if (l <= 0 || y >= 176)
+    return;
+
+  if (y & 1)
+    c = alternate(c >> 3, c);
+
+  int bitstart = x * 3;
+  int bitend = (x + l) * 3;
+  int wstart = bitstart >> 5;
+  int wend = bitend >> 5;
+  int padstart = bitstart & 31;
+  int padend = bitend & 31;
+  int maskstart = -1 << padstart;
+  int maskend = unsigned(-1) >> (32 - padend);
+  if (wstart == wend) {
+    maskstart &= maskend;
+    maskend = 0;
+  }
+
+  int* row = (int*) &fb[y * stride];
```

Comment:
Will this always be aligned to an `int` ? (can't remember if the chip we're on has strict alignment, so I'll just note this & close it, so if we do see bus errors etc, we can come back another time)

Reply:
It won't always be aligned, but the chip does support unaligned memory accesses so it's OK. :)

Reply:
Sounds grand, thanks for the quick reply too!